include/dxc/DxilContainer/DxilContainer.h

```diff
@@ -210,6 +212,141 @@ struct DxilShaderDebugName {
+//
+//         ...
+//
+//       DxilSourceInfoSection
+//       char Data[]
+//       (0-3 zero bytes to align to a 4-byte boundary)
+//
+// Each DxilSourceInfoSection is followed by a blob of Data.
+// The each type of data has its own internal structure:
+//
+// ================ 1. Sources ==================================
+// 
+//  DxilSourceInfo_Sources
+//    char Data[SizeInBytes]
+//   
+// `Data` may be compressed. When uncompressed, this is the structure:
+//
+//   DxilSourceInfo_SourcesElement
+//   char Name   [ NameSize ]    + NullTerminator
+//   char Content[ ContentSize ] + NullTerminator
+//   (0-3 zero bytes to align to a 4-byte boundary)
+//
+//   DxilSourceInfo_SourcesElement
+//   char Name   [ NameSize ]    + NullTerminator
+//   char Content[ ContentSize ] + NullTerminator
+//   (0-3 zero bytes to align to a 4-byte boundary)
+//
+//     ...
+//
+//   DxilSourceInfo_SourcesElement
+//   char Name   [ NameSize ]    + NullTerminator
+//   char Content[ ContentSize ] + NullTerminator
+//   (0-3 zero bytes to align to a 4-byte boundary)
+//
+// ================ 2. Defines ==================================
+//
+//   DxilSourceInfo_StringList
+//     char Define[] + NullTerminator
+//     char Define[] + NullTerminator
+//       ...
+//     char Define[] + NullTerminator
+//     NullTerminator (to mark no more options)
+//
+// ================ 3. Args ==================================
+//
+//   DxilSourceInfo_StringList
+//     char Arg[] + NullTerminator
+//     char Arg[] + NullTerminator
+//       ...
+//     char Arg[] + NullTerminator
+//     NullTerminator (to mark no more options)
+//
+// ================ 4. Target Profile ==================================
+//
+//   DxilSourceInfo_String
+//     char TargetProfile[] + NullTerminator
+//
+// ================ 5. Entry Point Name ==================================
+//
+//   DxilSourceInfo_String
+//     char EntryPoint[] + NullTerminator
+//
+
+struct DxilSourceInfo {
+  uint16_t Flags;         // Reserved, must be set to zero.
+  uint16_t SectionCount;  // The number of elements in the source info.
+  uint32_t SizeInDwords;
+};
+
+enum class DxilSourceInfoSectionType : uint16_t {
+  Sources,
+  Defines,
+  Args,
+  TargetProfile,
+  EntryPoint,
+};
+
+struct DxilSourceInfoSection {
+  uint16_t Flags;                   // Reserved, must be set to zero.
+  DxilSourceInfoSectionType Type;   // The type of data following this header.
+  uint32_t SizeInDwords;            // Size of the element, including this header, and the padding
+};
+
+struct DxilSourceInfo_StringList {
+  uint16_t Flags;       // Reserved, must be set to zero.
+  uint16_t SizeInBytes; // Length of all strings, including the double null terminator, not including this header.
+  uint32_t Count;
+};
+
+struct DxilSourceInfo_String {
+  uint16_t Flags;       // Reserved, must be set to zero.
+  uint16_t SizeInBytes; // Length of the string, not including null terminator
```

Comment:
It's a little odd that this size does not include null terminator, but the StringList case does include terminators in it's SizeInBytes.  Maybe we should just include the null terminator in the size here for consistency, and reduce chance of accidentally copying string without null termination?

Reply:
I guess a similar argument could be made for other string lengths as well (like NameSize, ContentSize in DxilSourceInfo_SourcesElement).

Reply:
With this kind of change, data format comments could be something like:
```
//   char Name   [ NameSize ];     // null-terminated
```
Which more closely reflects structure without having to infer that "+ NullTerminator" is equivalent to an extra field, like:
```
//   char Name   [ NameSize ];  char NullTerminator;
```
And without leaving open the question of whether NullTerminator is included if string is empty.

When including the null terminator in size, null and empty strings can be differentiated (if necessary), and it makes it clear how many bytes follow.

Reply:
I'm following the other structures in the container, like:
```
struct DxilShaderDebugName {
  uint16_t Flags;       // Reserved, must be set to zero.
  uint16_t NameLength;  // Length of the debug name, without null terminator.
  // Followed by NameLength bytes of the UTF-8-encoded name.
  // Followed by a null terminator.
  // Followed by [0-3] zero bytes to align to a 4-byte boundary.
};
```

Also I think it matches the way we think about strings in c, with strlen being the length without the terminator.
I think to make things more consistent, I'll just change the Size of null separated StringList to also not include the final terminator, or remove the double null terminator altogether.